### PR TITLE
[heft-config-file] Use an internal promise cache to reduce total file reads

### PR DESCRIPTION
**Pull request by elliot-nelson**
## Summary

Implements https://github.com/microsoft/rushstack/issues/2691.

By using an internal cache of _promises_ for Cache Entries, instead of Cache Entries themselves, we can reduce the total number of file system operations -- especially when rush is running parallel builds.

## Details

I tried implementing this in a variety of different ways, but modifying `_loadConfigurationFileInnerWithCacheAsync` gives the most bang for the buck.  This function is recursive, in that it may run off and load another config file due to an `extends` property, or hit a missing file and go and look for a rig'd version (both of which end up calling this same method again with different parameters).  Since this function now uses a _promise_ cache, that means that not only do two parallel attempts to load `a/rush-project.json` collapse to 1 file system read, but loading `a/rush-project.json` and `b/rush-project.json`, both of which extend `c/rush-project.json`, will collapse to 1 file system read for `c/rush-project.json`.

This has one negative side effect, which is that you can't effectively check for dependency loops in `_loadConfigurationFileInnerAsync` anymore.  I've added a comment to the function to that effect.

Now that entering the `_loadConfigurationFileInnerWithCacheAsync` funnel generates as few file system reads as possible, you want to enter the funnel as soon as possible -- it is counter-productive to do other reads up front (like the existing `existsSync` check).  I ended up ripping all that out, and simplifying the two entry point methods accordingly.

## How it was tested

There's a robust unit test suite already, which passed except for some minor logging behavior tweaks in some edge cases.

Validated that in our own monorepo, as expected, running `rush build` with these changes guarantees _exactly one_ load of any given configuration file in each pass (whereas before, if you had 5 concurrent builds, you'd see 5 attempts to load a common configuration file before it would start getting cache hits).

## Final thoughts

The only thing I still don't love about this approach is that there is one remaining known inefficiency -- handling missing files.

Today, all of the uses of this module are to call `tryLoadConfigurationFileForProjectAsync()`.  If a configuration file does not exist, we expect this to return `undefined`, but in order for that to happen:

1. `_loadConfigurationFileInnerWithCacheAsync` gets a cache hit with a `ENOENT` error, and then throws it.
2. `tryLoadConfigurationFileForProjectAsync` has to catch it, confirm it's an ENOENT, and return undefined.

It's still definitely a net win over making an extra 15 file system calls, but ideally, we would be able to route the knowledge that there is no file in the cache entry directly up to the end caller without having to throw a saved error and then cache it again.  I worked on a couple ideas but so far, I haven't struck upon an API change that doesn't break the other internal uses of `_loadConfigurationFileInnerWithCacheAsync`.

### CODE REVIEW

File: libraries/heft-config-file/src/test/ConfigurationFile.test.ts [hunk 1]
```diff
@@ -524,13 +524,9 @@ describe('ConfigurationFile', () => {
...
-      try {
-        expect(
-          await configFileLoader.tryLoadConfigurationFileForProjectAsync(terminal, __dirname)
-        ).toBeUndefined();
-      } catch (e) {
-        fail();
-      }
+      expect(
+        await configFileLoader.tryLoadConfigurationFileForProjectAsync(terminal, __dirname)
+      ).toBeUndefined();
```
Comment (elliot-nelson) [inline libraries/heft-config-file/src/test/ConfigurationFile.test.ts:529]: impl note: this change was not required for tests to pass, it's just a small enhancement to the test output if this test fails (show the exception you are getting rather than failing with no information).

**Comment by elliot-nelson** [comment]
Was still noodling on this issue and I think the thing to do, _if_ you cared about making cache hits as efficient as possible, would be to pay the cost in lines of code -- have two siblings internally:

```
loadConfigurationFileForProjectAsync -> _loadConfigurationFileInnerWithCacheAsync
tryLoadConfigurationFileForProjectAsync -> _tryLoadConfigurationFileInnerWithCacheAsync
```

The two `WithCache` methods are identical except that the `try` variant, if `cacheEntry.error` is defined, can check it on the spot for an ENOENT and return unknown, without having to go through the `throw->catch` path.  The other internal functions can call either internal `WithCache` variant, depending on their expected API.

(I don't know that we need to do this at all, but I think this would be the cleanest way.)

File: libraries/heft-config-file/src/ConfigurationFile.ts [hunk 2]
```diff
@@ -191,9 +191,10 @@ export class ConfigurationFile<TConfigurationFile> {
     return this.__schema;
   }
 
-  private readonly _configurationFileCache: Map<string, IConfigurationFileCacheEntry<TConfigurationFile>> =
-    new Map<string, IConfigurationFileCacheEntry<TConfigurationFile>>();
-  private readonly _fileExistsCache: Map<string, boolean> = new Map<string, boolean>();
+  private readonly _configPromiseCache: Map<
+    string,
+    Promise<IConfigurationFileCacheEntry<TConfigurationFile>>
+  > = new Map<string, Promise<IConfigurationFileCacheEntry<TConfigurationFile>>>();
```
Comment (dmichon-msft) [inline libraries/heft-config-file/src/ConfigurationFile.ts:197]: Pretty sure you should just be able to use `new Map()`; the type is already fully specified on the line, so I don't think you'll hit a lint error?

File: libraries/heft-config-file/src/ConfigurationFile.ts [hunk 3]
```diff
@@ -300,36 +286,61 @@ export class ConfigurationFile<TConfigurationFile> {
...
+          rigConfig
+        )
       );
+      this._configPromiseCache.set(resolvedConfigurationFilePath, cacheEntryPromise);
     }
 
+    // We check for loops after caching a promise for this config file, but before attempting
+    // to resolve the promise. We can't handle loop detection in the `InnerAsync` function, because
+    // we could end up waiting for a cached promise (like A -> B -> A) that never resolves.
+    if (visitedConfigurationFilePaths.has(resolvedConfigurationFilePath)) {
```
Comment (dmichon-msft) [inline libraries/heft-config-file/src/ConfigurationFile.ts:306]: This can be `else if`, since it cannot be in the visited array without a cache entry having already existed.

File: libraries/heft-config-file/src/ConfigurationFile.ts [hunk 4]
```diff
@@ -300,36 +286,61 @@ export class ConfigurationFile<TConfigurationFile> {
...
       return cacheEntry.configurationFile! as TConfigurationFile;
     }
   }
 
+  // Convert a promise for a TConfigurationFile into a promise for a corresponding
+  // cache entry instead.
+  private async _createCacheEntryPromise(
+    configFilePromise: Promise<TConfigurationFile>
+  ): Promise<IConfigurationFileCacheEntry<TConfigurationFile>> {
+    try {
```
Comment (dmichon-msft) [inline libraries/heft-config-file/src/ConfigurationFile.ts:330]: You should just be able to do:
```
configFilePromise.catch(() => { /* suppress unhandled rejection errors */ });
return configFilePromise;
```

**Review by dmichon-msft — COMMENTED** [review]
Couple of minor tweaks. I also think we ought to be able to somehow cache/deduplicate rig configs across workspace projects, since is is common to use the same Rig profile across a large number of projects, and that's a lot of redundant work.

**Comment by dmichon-msft** [comment]
I'm not worried about the overhead of `throw -> catch`; it's negligible compared to the file system operations.

File: libraries/heft-config-file/src/ConfigurationFile.ts (hunk 4 above)
Comment (elliot-nelson) [reply libraries/heft-config-file/src/ConfigurationFile.ts:330]: I don't think that would have the same API though.

This function converts one type of promise into a different type of promise:

 - The _input_ promise is a promise for a `TConfigurationFile`, and it will resolve if it exists and reject with an error if it does not.
 - The _output_ promise is a promise for a `IConfigurationFileCacheEntry`.  It should *resolve* with the object `{configurationFile: [thing]}` if the input promise resolves, and it should *resolve* with the object `{error: [thing]}` if the input promise rejects.

I don't think I can suppress any errors here, because I need to save that error in the cache entry.

File: libraries/heft-config-file/src/ConfigurationFile.ts (hunk 3 above)
Comment (elliot-nelson) [reply libraries/heft-config-file/src/ConfigurationFile.ts:306]: I agree with the logic, but readability-wise I'm not sure.

Right now the function is basically 3 snippets -- get or set a promise cache entry, perform loop detection, and then resolve or reject a value.  I see the "perform loop detection" snippet more as a safeguard for the last section (ensuring we don't accidentally wait on a promise that will never resolve), not so much related to the first snippet (get or set promise cache entry).

Turning it into an else-if merges the first and second snippet in a way that I think might be more confusing than helpful (but I guess it does save a function call in some situations)?

**Comment by elliot-nelson** [comment]
@dmichon-msft Thanks for reviewing this!  I made one fix and pushed back slightly on the other two, let me know what you think.

I also did look at `RigConfig` and you are spot on, it looks like there's no type of caching there at all yet (or, maybe a little bit of caching inside a RigConfig instance, but if multiple people ask for `RigConfig.loadForProject('blah')`, you get a new copy every time, so that internal caching doesn't help that much).

It looks like `rig-package` is intended to be more generic than `heft-config-file`, so I guess it wouldn't make sense for `RigConfig` to use this logic here; but we can probably implement similar caching in RigConfig.

File: libraries/heft-config-file/src/ConfigurationFile.ts (hunk 3 above)
Comment (dmichon-msft) [reply libraries/heft-config-file/src/ConfigurationFile.ts:309]: It saves a lookup on the first read of every file. Granted, the lookup is a very minor cost relative to the file system reads, so it's a pretty minor performance gain.

File: libraries/heft-config-file/src/ConfigurationFile.ts (hunk 4 above)
Comment (dmichon-msft) [reply libraries/heft-config-file/src/ConfigurationFile.ts:333]: I meant to imply changing the promise type to `TConfigurationFile`, since a rejected promise can be reused to save the error just as easily as caching it in an object.

Comment (elliot-nelson) [reply libraries/heft-config-file/src/ConfigurationFile.ts:333]: Ohhh! Remove `Promise<IConfigurationFileCacheEntry>` completely and replace with `Promise< TConfigurationFile>`!

Makes total sense, I'll try that out and push it up if I don't run into any issues.

File: libraries/heft-config-file/src/ConfigurationFile.ts [hunk 5]
```diff
@@ -206,6 +207,11 @@ export class ConfigurationFile<TConfigurationFile> {
     this._propertyInheritanceTypes = options.propertyInheritance || {};
   }
 
+  /**
+   * Find and return a configuration file for the specified project, automatically resolving
+   * `extends` properties and handling rig'd configuration files. Will throw an error if a configuration
```
Comment (octogonz) [inline libraries/heft-config-file/src/ConfigurationFile.ts:212]:
```suggestion
   * `extends` properties and handling rigged configuration files. Will throw an error if a configuration
```

**Comment by octogonz** [comment]
> 
> I'm not worried about the overhead of `throw -> catch`; it's negligible compared to the file system operations.

I'm somewhat worried about "unexceptional exceptions." When a project throws a firehose of exceptions during normal successful operation, this can make the control flow difficult to analyze.  It completely ruins the VS Code break-on-exception feature.

However, the Node.js API basically forces us into this approach if we want to optimize performance.  (It would be better if they provided an API that could a report file-not-found result without throwing an exception.)

File: libraries/heft-config-file/src/ConfigurationFile.ts (hunk 2 above)
Comment (octogonz) [reply libraries/heft-config-file/src/ConfigurationFile.ts]: +1  This is more readable and completely cool in the type system:

`_configPromiseCache: Map<string, Promise<IConfigurationFileCacheEntry<TConfigurationFile>>> = new Map();`

**Comment by octogonz** [comment]
Thanks for implementing this!

LGTM, with David's suggestions

**Comment by elliot-nelson** [comment]
Latest commits include:

 - David's suggestion to eliminate the (no longer necessary) CacheEntry wrapper.
 - Pete's suggested comment change.

**Review by dmichon-msft — APPROVED** [review]

**Review by octogonz — APPROVED** [review]
